File: runtime/ts/src/dispatcher.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import type { Endpoint, Message, Receiver } from '@nightseam/duplex';
import { createDispatcher } from './dispatcher.ts';
import { wirePair } from './wire-pair.ts';
import { callWire, handleWire } from './wire.ts';

test('one dispatcher attachment supports siblings and leaves its borrowed endpoint usable', async () => {
  const [left, right] = wirePair();
  try {
    const dispatch = createDispatcher(right);
    assert.throws(() => right.receive({}), { code: 'receiver_exists' });
    for (const name of ['a', 'b']) {
      const binding = createDispatcher(dispatch.select([name]));
      handleWire(binding, ['read'], () => name);
    }
    for (const name of ['a', 'b']) assert.equal(await callWire(left, [name, 'read']), name);
    dispatch.close();
    const rebound = createDispatcher(right);
    handleWire(rebound, ['read'], () => 'rebound');
    assert.equal(await callWire(left, ['read']), 'rebound');
  } finally {
    left.close();
  }
});

test('an unmanaged invocation is explicitly refused with its original return capability', () => {
  let attached: Receiver | undefined;
  const endpoint: Endpoint = {
    send() {},
    receive(receiver) {
      attached = receiver;
      return () => {};
    },
    close() {},
  };
  const dispatch = createDispatcher(endpoint);
  let called = false;
  dispatch.register([], {
    message() {
      called = true;
    },
  });
  let reply: Message | undefined;
  // A return capability refuses what it does not implement, as every addressed
  // receiver in this profile does; this one carries outcomes and nothing else.
  const address = {
    wire: {
      send(path: readonly string[], message: Message) {
        if (path.length) throw new Error('this return capability carries outcomes only');
        reply = message;
      },
    },
  };
  attached!.message!([], { frame: { version: 1, kind: 'request', id: 'one', params: null }, return: address });
  assert.equal(called, false);
  assert.equal(reply?.frame.kind, 'response');
  if (reply?.frame.kind === 'response') assert.equal(reply.frame.error?.code, 'invalid_message');
});

test('an explicitly owned dispatcher closes its endpoint', () => {
  const [left, right] = wirePair();
  try {
    const dispatcher = createDispatcher(right, { ownEndpoint: true });
    dispatcher.close(1002, 'wire event rejected');
    assert.throws(() => right.receive({}), { code: 'disconnected' });
    dispatcher.close();
  } finally {
    left.close();
  }
});

File: runtime/go/dispatcher_test.go
package runtime_test

import (
	"context"
	"encoding/json"
	"testing"

	"github.com/Bitspark/nightseam/duplex/go"
	ws "github.com/Bitspark/nightseam/runtime/go"
)

func TestDispatcherSharesOneAttachmentAndPreservesBorrowedEndpoint(t *testing.T) {
	left, right, err := ws.NewWirePair(ws.Options{})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = left.Close(duplex.CodeNormal, "") })
	dispatch, err := ws.NewDispatcher(right)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := right.Receive(duplex.Receiver{}); err == nil {
		t.Fatal("second owning attachment accepted")
	}
	for _, name := range []string{"a", "b"} {
		view := dispatch.Select([]string{name})
		binding, err := ws.NewDispatcher(view)
		if err != nil {
			t.Fatal(err)
		}
		if _, err := ws.HandleWire(binding, []string{"read"}, func(context.Context, json.RawMessage) (any, error) { return name, nil }); err != nil {
			t.Fatal(err)
		}
	}
	for _, name := range []string{"a", "b"} {
		var got string
		if err := ws.CallWire(context.Background(), left, []string{name, "read"}, nil, &got); err != nil || got != name {
			t.Fatalf("%s: %q, %v", name, got, err)
		}
	}
	if err := dispatch.Close(duplex.CodeNormal, ""); err != nil {
		t.Fatal(err)
	}
	rebound, err := ws.NewDispatcher(right)
	if err != nil {
		t.Fatalf("borrowed endpoint was closed: %v", err)
	}
	if _, err := ws.HandleWire(rebound, []string{"read"}, func(context.Context, json.RawMessage) (any, error) { return "rebound", nil }); err != nil {
		t.Fatal(err)
	}
	var got string
	if err := ws.CallWire(context.Background(), left, []string{"read"}, nil, &got); err != nil || got != "rebound" {
		t.Fatalf("rebound: %q, %v", got, err)
	}
}

func TestDispatcherClosesAnExplicitlyOwnedEndpoint(t *testing.T) {
	left, right, err := ws.NewWirePair(ws.Options{})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = left.Close(duplex.CodeNormal, "") })
	dispatch, err := ws.NewDispatcher(right, ws.DispatcherOptions{OwnEndpoint: true})
	if err != nil {
		t.Fatal(err)
	}
	if err := dispatch.Close(duplex.CodeProtocolError, "wire event rejected"); err != nil {
		t.Fatal(err)
	}
	if _, err := right.Receive(duplex.Receiver{}); err == nil {
		t.Fatal("owned endpoint stayed open")
	}
	if err := dispatch.Close(duplex.CodeNormal, "again"); err != nil {
		t.Fatal(err)
	}
}

func TestDispatcherRefusesUnmanagedInvocationWithoutReplacingReturn(t *testing.T) {
	root := &unmanagedDispatchEndpoint{}
	dispatch, err := ws.NewDispatcher(root)
	if err != nil {
		t.Fatal(err)
	}
	called := false
	if _, err := dispatch.Register(nil, duplex.Receiver{Message: func([]string, duplex.Message) { called = true }}); err != nil {
		t.Fatal(err)
	}
	replies := &wireReplySink{replies: make(chan duplex.ProfileFrame, 1)}
	address := &duplex.ReturnAddress{Wire: replies}
	root.receiver.Message(nil, duplex.Message{Frame: duplex.ProfileFrame{Version: 1, Kind: duplex.ProfileRequest, ID: "one", Params: json.RawMessage(`null`)}, Return: address})
	if called {
		t.Fatal("unmanaged invocation dispatched without a lifetime association")
	}
	if reply := receive(t, replies.replies); reply.Error == nil || reply.Error.Code != "invalid_message" {
		t.Fatalf("unmanaged refusal = %+v", reply)
	}
}

type unmanagedDispatchEndpoint struct{ receiver duplex.Receiver }

func (*unmanagedDispatchEndpoint) Send([]string, duplex.Message) error { return nil }
func (w *unmanagedDispatchEndpoint) Receive(receiver duplex.Receiver) (func(), error) {
	w.receiver = receiver
	return func() {}, nil
}
func (*unmanagedDispatchEndpoint) Close(duplex.Code, string) error { return nil }
